storage/merge.go

```diff
@@ -458,47 +462,59 @@ func (c *chainSampleIterator) Seek(t int64) bool {
 			heap.Push(&c.h, iter)
 		}
 	}
-	return len(c.h) > 0
+	if len(c.h) > 0 {
+		c.curr = heap.Pop(&c.h).(chunkenc.Iterator)
+		return true
+	}
+	return false
 }
 
 func (c *chainSampleIterator) At() (t int64, v float64) {
-	if len(c.h) == 0 {
-		panic("chainSampleIterator.At() called after .Next() returned false.")
+	if c.h == nil {
+		panic("chainSampleIterator.At() called before first .Next()")
 	}
-
-	return c.h[0].At()
+	return c.curr.At()
 }
 
 func (c *chainSampleIterator) Next() bool {
 	if c.h == nil {
-		for _, iter := range c.iterators {
+		c.h = samplesIteratorHeap{}
+		c.curr = c.iterators[0]
+		for _, iter := range c.iterators[1:] {
 			if iter.Next() {
 				heap.Push(&c.h, iter)
 			}
 		}
-
-		return len(c.h) > 0
 	}
 
-	if len(c.h) == 0 {
-		return false
-	}
+	for {
+		if c.curr.Next() {
+			currt, _ := c.curr.At()
+			if currt == c.lastt {
```

Comment:
All should work, but I am not huge fan of checking duplicates within a chunk/single iterator itself :thinking: 

Let me benchmark it if it has any overhead...

Reply:
No overhead.